db/system_keyspace.cc

```diff
@@ -3228,6 +3228,56 @@ future<mutation> system_keyspace::get_group0_history(distributed<service::storag
     co_return mutation(s, partition_key::from_singular(*s, GROUP0_HISTORY_KEY));
 }
 
+static constexpr auto GROUP0_UPGRADE_STATE_KEY = "group0_upgrade_state";
+
+future<service::group0_upgrade_state> system_keyspace::load_group0_upgrade_state() {
+    auto s = co_await get_scylla_local_param_as<sstring>(GROUP0_UPGRADE_STATE_KEY);
+
+    if (!s || *s == "use_old_procedures") {
+        co_return service::group0_upgrade_state::use_old_procedures;
+    } else if (*s == "synchronize") {
+        co_return service::group0_upgrade_state::synchronize;
+    } else if (*s == "use_new_procedures") {
+        co_return service::group0_upgrade_state::use_new_procedures;
+    } else if (*s == "recovery") {
+        co_return service::group0_upgrade_state::recovery;
+    }
+
+    slogger.error(
+            "load_group0_upgrade_state(): unknown value '{}' for key 'group0_upgrade_state' in Scylla local table."
+            " Did you change the value manually?"
+            " Correct values are: 'use_old_procedures', 'synchronize', 'use_new_procedures', 'recovery'."
+            " Assuming 'recovery'.", *s);
+    // We don't call `on_internal_error` which would probably prevent the node from starting, but enter `recovery`
+    // allowing the user to fix their cluster.
+    co_return service::group0_upgrade_state::recovery;
+}
+
+future<> system_keyspace::save_group0_upgrade_state(service::group0_upgrade_state s) {
+    auto value = [s] () -> sstring {
+        switch (s) {
+            case service::group0_upgrade_state::use_new_procedures:
```

Comment:
Can't  C++ help you with some compiler magic here so that you don't have to have this switch? There is even https://github.com/ToniBig/cpp-named-enum and https://github.com/Neargye/magic_enum for that. If you decide to write it by hand, I think you should at least add constexpr.

Reply:
Another simple alternative is to not store a string but a simple integer, or, not use an enum, but a simple string.